streaming/src/main/scala/org/apache/spark/streaming/receiver/BlockGenerator.scala

```diff
@@ -140,6 +140,16 @@ private[streaming] class BlockGenerator(
       }
     }
 
+    // When we arrive here, no data will be added to `currentBuffer`. However, `currentBuffer` may
+    // not be empty. If so, we should wait until all data in `currentBuffer` is consumed, because
+    // `blockIntervalTimer.stop(interruptTimer = false)` doesn't guarantee calling
+    // `updateCurrentBuffer` for us.
+    var isCurrentBufferEmpty = synchronized { currentBuffer.isEmpty }
+    while(!isCurrentBufferEmpty) {
+      Thread.sleep(blockIntervalMs)
+      isCurrentBufferEmpty = synchronized { currentBuffer.isEmpty }
```

Comment:
A better fix would be to ensure that `updateCurrentBuffer` is called when `blockIntervalTimer` is called. This can be done by simply adding a parameter to `RecurringTimer` to ensure that the callback is made in the `stop` method. 

The advantage is that we won't be burning CPU here.